src/nfa/intersection.cc

```diff
@@ -16,177 +16,233 @@
 // MATA headers
 #include "mata/nfa/nfa.hh"
 #include "mata/nfa/algorithms.hh"
+#include <cassert>
+#include <functional>
+
 
 using namespace mata::nfa;
 
 namespace {
 
-/**
- * Add transition to the product.
- * @param[out] product Created product automaton.
- * @param[out] product_map Created product map.
- * @param[in] pair_to_process Currently processed pair of original states.
- * @param[in] intersection_transition State transitions to add to the product.
- */
-void add_product_transition(Nfa& product, std::unordered_map<std::pair<State,State>, State>& product_map,
-                            const std::pair<State,State>& pair_to_process, SymbolPost& intersection_transition) {
-    if (intersection_transition.empty()) { return; }
-
-    auto& intersect_state_transitions{ product.delta.mutable_state_post(product_map[pair_to_process]) };
-    auto intersection_move_iter{ intersect_state_transitions.find(intersection_transition) };
-    if (intersection_move_iter == intersect_state_transitions.end()) {
-        intersect_state_transitions.insert(intersection_transition);
-    } else {
-        // Product already has some target states for the given symbol from the current product state.
-        intersection_move_iter->insert(intersection_transition.targets);
-    }
+using ProductMap = std::unordered_map<std::pair<State,State>,State>;
+using MatrixProductStorage = std::vector<std::vector<State>>;
+using VecMapProductStorage = std::vector<std::unordered_map<State,State>>;
+using InvertedProductStorage = std::vector<State>;
+//Unordered map seems to be faster than ordered map here, but still very much slower than matrix.
+
+} // Anonymous namespace.
+
+namespace mata::nfa {
+
+Nfa intersection(const Nfa& lhs, const Nfa& rhs, const Symbol first_epsilon, ProductMap *prod_map) {
+
```

Comment:
Btw. one small suggestion for potential speedup: test the following:
```
if lhs.empty:
   return lhs
elif rhs.empty:
  return rhs
```
Might show some speed up, if in the product you do some initialization steps.

Reply:
Might make sense, but I am not sure whether emptiness is really for free, we are changing it now, lets see the performance first.

Reply:
As I said at meeting, I don't suggest testing language emptiness, but maybe delta emptiness? Or final state emptiness could be applied? Basically anything cheap, that can be used to avoid costly initializations.

Reply:
aha, ok

Reply:
fine, I added test for final initial state emptiness